File: backend/src/app/schemas/onboarding_oversight.py
import uuid as uuid_pkg
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from .rp_application_promotion_request import ProductionReviewStatus


class ProductionReviewQueueRowRead(BaseModel):
    """Cross-workspace CL Admin view of one explicit Production review."""

    model_config = ConfigDict(
        extra="forbid",
        validate_by_name=True,
        validate_by_alias=True,
        alias_generator=to_camel,
        populate_by_name=True,
    )

    rp_configuration_uuid: uuid_pkg.UUID
    configuration_name: str = Field(..., min_length=1, max_length=128)
    source_rp_configuration_uuid: uuid_pkg.UUID | None = None
    application_information_uuid: uuid_pkg.UUID
    application_name_en: str = Field(..., min_length=1, max_length=255)
    application_name_fr: str = Field(..., min_length=1, max_length=255)
    workspace_uuid: uuid_pkg.UUID
    workspace_name: str = Field(..., min_length=1, max_length=128)
    department_uuid: uuid_pkg.UUID | None = None
    department_name: str | None = Field(default=None, max_length=100)
    review_status: ProductionReviewStatus
    external_review_reference: str = Field(..., min_length=1, max_length=255)
    reviewed_by_user_uuid: uuid_pkg.UUID | None = None
    reviewed_by_team: str | None = Field(default=None, max_length=128)
    requested_at: datetime
    reviewed_at: datetime | None = None
    decided_at: datetime | None = None
    updated_at: datetime | None = None
    detail_path: str = Field(..., min_length=1)


# Internal import compatibility; the serialized contract no longer contains
# an onboarding record type or lifecycle state.
OnboardingOversightQueueRowRead = ProductionReviewQueueRowRead

File: frontend/src/fetch/onboarding-oversight.ts
import { requestJson } from "./request-json";
import {
	normalizeOnboardingOversightQueueFilters,
	type OnboardingOversightQueueFilters,
	type OnboardingOversightQueueReviewStatus,
} from "@/features/onboarding-oversight/queue-filters";

export type OnboardingOversightQueueRowRead = {
	applicationInformationUuid: string;
	applicationNameEn: string;
	applicationNameFr: string;
	configurationName: string;
	departmentName?: string | null;
	departmentUuid?: string | null;
	detailPath: string;
	decidedAt?: string | null;
	externalReviewReference: string;
	requestedAt: string;
	reviewedAt?: string | null;
	reviewedByTeam?: string | null;
	reviewedByUserUuid?: string | null;
	reviewStatus: OnboardingOversightQueueReviewStatus;
	rpConfigurationUuid: string;
	sourceRpConfigurationUuid?: string | null;
	updatedAt?: string | null;
	workspaceName: string;
	workspaceUuid: string;
};

const buildQueryString = (filters: OnboardingOversightQueueFilters): string => {
	const normalizedFilters = normalizeOnboardingOversightQueueFilters(filters);
	const params = new URLSearchParams();

	if (normalizedFilters.department) {
		params.set("department", normalizedFilters.department);
	}
	if (normalizedFilters.reviewStatus) {
		params.set("review_status", normalizedFilters.reviewStatus);
	}
	if (normalizedFilters.workspace) {
		params.set("workspace", normalizedFilters.workspace);
	}

	const queryString = params.toString();
	return queryString ? `?${queryString}` : "";
};

export const getOnboardingOversightQueue = async (
	filters: OnboardingOversightQueueFilters = {}
): Promise<Array<OnboardingOversightQueueRowRead>> => {
	const result =
		await requestJson<Array<OnboardingOversightQueueRowRead> | null>(
			`/api/v1/onboarding-oversight/production-reviews${buildQueryString(filters)}`,
			{
				cache: "no-store",
				method: "GET",
			}
		);

	return result ?? [];
};
